src/crimson/os/seastore/cache.cc

```diff
@@ -1024,6 +1028,15 @@ CachedExtentRef Cache::duplicate_for_write(
   return ret;
 }
 
+void Cache::set_segment_provider(AsyncCleaner *cleaner) {
+  auto provider = dynamic_cast<SegmentProvider*>(cleaner);
```

Comment:
I'd like to avoid using dynamic_cast.  Can we add a maybe_get_provider() method to the AsyncCleaner interface?

Reply:
Probably maybe_get_segment_provider() as it is segment-specific.

Reply:
How about constructing `std::vector<SegmentProvider*>` outside the Cache. In `TransactionManager::make_transaction_manager` we cast cast `SegmentCleaner*` to `SegmentProvider*` safely.

Reply:
Yeah, the point is to use static-cast outside.